### PR TITLE
[Bug Fix] Invis vs. Undead/Animal Breaks Charm for Pets

### PR DESCRIPTION
**Pull request by crdunwel**
Fixes https://github.com/EQEmu/Server/issues/2212

### CODE REVIEW

File: zone/mob.cpp [hunk 1]
```diff
@@ -593,57 +593,48 @@ void Mob::CalcSeeInvisibleLevel()
 	see_invis = std::max({ spellbonuses.SeeInvis, itembonuses.SeeInvis, aabonuses.SeeInvis, innate_see_invis });
 }
 
-void Mob::CalcInvisibleLevel()
-{
-	bool is_invisible = invisible;
+bool Mob::HasAnInvisibilityEffect() {
```
Comment (Akkadius) [inline zone/mob.cpp:596]: Love it

File: zone/mob.cpp [hunk 2]
```diff
@@ -593,57 +593,48 @@ void Mob::CalcSeeInvisibleLevel()
...
+	return invisible || hidden || improved_hidden || invisible_animals || invisible_undead;
+}
+
+void Mob::BreakCharmPetIfConditionsMet() {
+	auto pet = GetPet();
+	if (pet && pet->GetPetType() == petCharmed && HasAnInvisibilityEffect()) {
+		if (RuleB(Pets, LivelikeBreakCharmOnInvis) || IsInvisible(pet)) {
+			pet->BuffFadeByEffect(SE_Charm);
+		}
+		LogRules("Pets:LivelikeBreakCharmOnInvis for [{}] | Invis [{}] - Hidden [{}] - Shroud of Stealth [{}] - IVA [{}] - IVU [{}]",
```
Comment (Akkadius) [inline zone/mob.cpp]: Adjusted formatting

```cpp
LogRules(
	"Pets:LivelikeBreakCharmOnInvis for [{}] invisible [{}] hidden [{}] improved_hidden (shroud of stealth) [{}] invisible_animals [{}] invisible_undead [{}]",
	GetCleanName(),
	invisible,
	hidden,
	improved_hidden,
	invisible_animals,
	invisible_undead
);
```

File: zone/mob.cpp [hunk 3]
```diff
@@ -593,57 +593,48 @@ void Mob::CalcSeeInvisibleLevel()
...
 	invisible_undead = spellbonuses.invisibility_verse_undead;
 	invisible_animals = spellbonuses.invisibility_verse_animal;
 
-	if (!is_invisible && invisible) {
-		SetInvisible(Invisibility::Invisible, true);
-		return;
-	}
-
-	if (is_invisible && !invisible) {
+	if (was_invisible != invisible) {
```
Comment (Akkadius) [inline zone/mob.cpp:626]: Adjusted formatting here

Try to not use `else` if we can. Prevents code getting needlessly nested in the long term of maintenance.

Adjusted to

```cpp
void Mob::CalcInvisibleLevel()
{
	bool was_invisible = invisible;

	invisible         = std::max({spellbonuses.invisibility, nobuff_invisible});
	invisible_undead  = spellbonuses.invisibility_verse_undead;
	invisible_animals = spellbonuses.invisibility_verse_animal;

	if (was_invisible != invisible) {
		SetInvisible(invisible, true);
		return;
	}

	BreakCharmPetIfConditionsMet();
}
```

**Comment by Akkadius** [comment]
Looks good to me, can someone validate functionality here. @noudess you've done work with charmed pets

**Comment by noudess** [comment]
> Looks good to me, can someone validate functionality here. @noudess you've done work with charmed pets

I love the refactoring as it makes the code more readable, but I don't see how the functioanlity is altered.  Can you explain what is new/different that makes it work?

Ah.. was it this change:

if (is_invisible && !invisible) {
to:
if (was_invisible != invisible) {

**Review by noudess — APPROVED** [review]
The fact that CalcInvisibleLevel() never called the function that checked for IVU and IVA is astonishing.  Pulling that up into the function is a good solution.  Also, the check on the previous and new invis state is more accurate comparing them to each other.  The old code relied on code evsewhere to detect the change from visible -> invisible.

Nice Job.

**Pull request merged**